In [23]:
import pandas as pd
import numpy as np
from IPython.display import display

In [24]:

datasets = {
    "hotel_bookings": "../data/raw/hotel_bookings.csv",
    "customers": "../data/raw/olist_customers_dataset.csv",
    "geolocation": "../data/raw/olist_geolocation_dataset.csv",
    "order_items": "../data/raw/olist_order_items_dataset.csv",
    "payments": "../data/raw/olist_order_payments_dataset.csv",
    "reviews": "../data/raw/olist_order_reviews_dataset.csv",
    "orders": "../data/raw/olist_orders_dataset.csv",
    "products": "../data/raw/olist_products_dataset.csv",
    "sellers": "../data/raw/olist_sellers_dataset.csv",
    "category_translation": "../data/raw/product_category_name_translation.csv"
}



dfs = {}

for name, path in datasets.items():
    dfs[name] = pd.read_csv(path)

def analyze_dataset(name, df):

    print("\n" + "="*120)
    print(f"DATASET: {name.upper()}")
    print("="*120)


    print("\n[1] SAMPLE DATA (10 ROWS)")

    sample_size = min(10, len(df))
    display(df.sample(sample_size))


    print("\n[2] DATA SHAPE")

    shape_df = pd.DataFrame({
        "Rows": [df.shape[0]],
        "Columns": [df.shape[1]]
    })

    display(shape_df)


    print("\n[3] DATA TYPES")

    dtype_df = pd.DataFrame({
        "Column": df.columns,
        "Data Type": df.dtypes.astype(str)
    })

    display(dtype_df)


    print("\n[4] DESCRIPTIVE STATISTICS")

    display(df.describe(include="all").T)

    print("\n[5] MISSING VALUES")

    missing_df = pd.DataFrame({
        "Missing Count": df.isnull().sum(),
        "Missing Percentage (%)":
            round(df.isnull().mean() * 100, 2)
    })

    missing_df = (
        missing_df
        .sort_values(
            by="Missing Percentage (%)",
            ascending=False
        )
    )

    display(missing_df)

    print("\n[6] DUPLICATED ROWS")

    duplicate_count = df.duplicated().sum()

    duplicate_df = pd.DataFrame({
        "Duplicate Rows": [duplicate_count],
        "Duplicate Percentage (%)": [
            round(
                duplicate_count / len(df) * 100,
                2
            )
        ]
    })

    display(duplicate_df)


    print("\n[7] CARDINALITY")

    cardinality_df = pd.DataFrame({
        "Unique Values": df.nunique()
    })

    cardinality_df = (
        cardinality_df
        .sort_values(
            by="Unique Values",
            ascending=False
        )
    )

    display(cardinality_df)


for name, df in dfs.items():
    analyze_dataset(name, df)


DATASET: HOTEL_BOOKINGS

[1] SAMPLE DATA (10 ROWS)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
7157,Resort Hotel,0,284,2016,July,31,24,2,3,2,...,No Deposit,NaN,NaN,0,Transient,169.50,1,1,Check-Out,2016-07-29
106549,City Hotel,0,134,2017,February,8,25,0,1,1,...,No Deposit,30.0,NaN,0,Transient-Party,75.00,0,0,Check-Out,2017-02-26
35213,Resort Hotel,0,11,2017,April,15,11,0,5,2,...,No Deposit,40.0,NaN,0,Contract,82.85,0,1,Check-Out,2017-04-16
87353,City Hotel,0,53,2016,April,16,13,1,4,2,...,No Deposit,14.0,NaN,0,Transient,140.75,1,0,Check-Out,2016-04-18
110300,City Hotel,0,65,2017,April,17,23,2,0,2,...,No Deposit,7.0,NaN,0,Transient,83.16,0,0,Check-Out,2017-04-25
44349,City Hotel,0,61,2015,October,41,4,2,0,1,...,No Deposit,12.0,NaN,0,Transient-Party,73.00,0,0,Check-Out,2015-10-06
73527,City Hotel,1,159,2017,August,34,25,0,2,2,...,No Deposit,9.0,NaN,0,Transient,139.50,0,0,Canceled,2017-03-21
97895,City Hotel,0,159,2016,September,39,19,1,2,2,...,No Deposit,240.0,NaN,0,Transient-Party,144.00,0,0,Check-Out,2016-09-22
45323,City Hotel,0,69,2015,October,44,25,2,1,1,...,No Deposit,37.0,NaN,58,Transient-Party,75.33,0,0,Check-Out,2015-10-28
34763,Resort Hotel,0,26,2017,April,14,3,1,0,2,...,No Deposit,NaN,251.0,0,Transient,62.00,0,0,Check-Out,2017-04-04



[2] DATA SHAPE


,Rows,Columns
0,119390,32



[3] DATA TYPES


,Column,Data Type
hotel,hotel,object
is_canceled,is_canceled,int64
lead_time,lead_time,int64
arrival_date_year,arrival_date_year,int64
arrival_date_month,arrival_date_month,object
arrival_date_week_number,arrival_date_week_number,int64
arrival_date_day_of_month,arrival_date_day_of_month,int64
stays_in_weekend_nights,stays_in_weekend_nights,int64
stays_in_week_nights,stays_in_week_nights,int64
adults,adults,int64



[4] DESCRIPTIVE STATISTICS


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
hotel,119390,2,City Hotel,79330,NaN,NaN,NaN,NaN,NaN,NaN,NaN
is_canceled,119390.0,NaN,NaN,NaN,0.370416,0.482918,0.0,0.0,0.0,1.0,1.0
lead_time,119390.0,NaN,NaN,NaN,104.011416,106.863097,0.0,18.0,69.0,160.0,737.0
arrival_date_year,119390.0,NaN,NaN,NaN,2016.156554,0.707476,2015.0,2016.0,2016.0,2017.0,2017.0
arrival_date_month,119390,12,August,13877,NaN,NaN,NaN,NaN,NaN,NaN,NaN
arrival_date_week_number,119390.0,NaN,NaN,NaN,27.165173,13.605138,1.0,16.0,28.0,38.0,53.0
arrival_date_day_of_month,119390.0,NaN,NaN,NaN,15.798241,8.780829,1.0,8.0,16.0,23.0,31.0
stays_in_weekend_nights,119390.0,NaN,NaN,NaN,0.927599,0.998613,0.0,0.0,1.0,2.0,19.0
stays_in_week_nights,119390.0,NaN,NaN,NaN,2.500302,1.908286,0.0,1.0,2.0,3.0,50.0
adults,119390.0,NaN,NaN,NaN,1.856403,0.579261,0.0,2.0,2.0,2.0,55.0



[5] MISSING VALUES


,Missing Count,Missing Percentage (%)
company,112593,94.31
agent,16340,13.69
country,488,0.41
hotel,0,0.00
arrival_date_month,0,0.00
arrival_date_week_number,0,0.00
lead_time,0,0.00
is_canceled,0,0.00
stays_in_weekend_nights,0,0.00
stays_in_week_nights,0,0.00



[6] DUPLICATED ROWS


,Duplicate Rows,Duplicate Percentage (%)
0,31994,26.8



[7] CARDINALITY


,Unique Values
adr,8879
reservation_status_date,926
lead_time,479
company,352
agent,333
country,177
days_in_waiting_list,128
previous_bookings_not_canceled,73
arrival_date_week_number,53
stays_in_week_nights,35



DATASET: CUSTOMERS

[1] SAMPLE DATA (10 ROWS)


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
43263,73038e9f26a27fc56631416306bd7829,c3ddcb8cafa47711e96ff4852fc7e52c,11703,praia grande,SP
83058,e405219e7366a91eb22f74d3926415d6,2ba34389582f8a27e6e9bdde979cb2ef,14085,ribeirao preto,SP
90065,abf36d4666c68c17927c23c39364fac4,f108981245a86b509868e5fc57a3a85c,4844,sao paulo,SP
52105,2e42725b340b879a9e4e93caac411a3e,94e416a11c3c2f3f0ec534356e4117f6,6363,carapicuiba,SP
60976,cf7ce922150eda8741cc086fa1ba721b,3d54514627343e42a4e59c88d68df5f9,83324,pinhais,PR
94846,82ad287ffd79d1492f0291e8d41dde20,0b6e816bf951b83bb07f8c97adb3c1ae,3968,sao paulo,SP
54235,04872433c17e4ca13b0aa2381838d11e,b3e529a2053eaec39876a97f990c3726,36576,vicosa,MG
67602,b5fafd3b606b479ad14ecad6eeeade71,5a46ee8cfcbdc029db7152e5ce3faeb1,95703,bento goncalves,RS
43957,8f255887fe8b85c220e25b63ccd30a06,cd6dfb23b4b61776cbf3ab819b69c108,2333,sao paulo,SP
68852,569ca092bc9e14268fcf7a5448de8647,abe6e71cf427fdd91fb41fa6a6cd2fb8,4302,sao paulo,SP



[2] DATA SHAPE


,Rows,Columns
0,99441,5



[3] DATA TYPES


,Column,Data Type
customer_id,customer_id,object
customer_unique_id,customer_unique_id,object
customer_zip_code_prefix,customer_zip_code_prefix,int64
customer_city,customer_city,object
customer_state,customer_state,object



[4] DESCRIPTIVE STATISTICS


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
customer_id,99441,99441,274fa6071e5e17fe303b9748641082c8,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
customer_unique_id,99441,96096,8d50f5eadf50201ccdcedfb9e2ac8455,17,NaN,NaN,NaN,NaN,NaN,NaN,NaN
customer_zip_code_prefix,99441.0,NaN,NaN,NaN,35137.474583,29797.938996,1003.0,11347.0,24416.0,58900.0,99990.0
customer_city,99441,4119,sao paulo,15540,NaN,NaN,NaN,NaN,NaN,NaN,NaN
customer_state,99441,27,SP,41746,NaN,NaN,NaN,NaN,NaN,NaN,NaN



[5] MISSING VALUES


,Missing Count,Missing Percentage (%)
customer_id,0,0.0
customer_unique_id,0,0.0
customer_zip_code_prefix,0,0.0
customer_city,0,0.0
customer_state,0,0.0



[6] DUPLICATED ROWS


,Duplicate Rows,Duplicate Percentage (%)
0,0,0.0



[7] CARDINALITY


,Unique Values
customer_id,99441
customer_unique_id,96096
customer_zip_code_prefix,14994
customer_city,4119
customer_state,27



DATASET: GEOLOCATION

[1] SAMPLE DATA (10 ROWS)


,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
600735,35670,-19.991719,-44.431770,mateus leme,MG
405043,20071,-22.901305,-43.180109,rio de janeiro,RJ
660911,39404,-16.719087,-43.809742,montes claros,MG
246378,11055,-23.968437,-46.332530,santos,SP
916167,88701,-28.487137,-49.013434,tubarao,SC
232564,9540,-23.621707,-46.565750,sao caetano do sul,SP
518388,28623,-22.299020,-42.536398,nova friburgo,RJ
224802,9280,-23.641268,-46.509244,santo andre,SP
889074,86450,-23.564877,-49.892430,quatigua,PR
754620,64660,-6.833888,-40.611877,pio ix,PI



[2] DATA SHAPE


,Rows,Columns
0,1000163,5



[3] DATA TYPES


,Column,Data Type
geolocation_zip_code_prefix,geolocation_zip_code_prefix,int64
geolocation_lat,geolocation_lat,float64
geolocation_lng,geolocation_lng,float64
geolocation_city,geolocation_city,object
geolocation_state,geolocation_state,object



[4] DESCRIPTIVE STATISTICS


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
geolocation_zip_code_prefix,1000163.0,NaN,NaN,NaN,36574.166466,30549.33571,1001.0,11075.0,26530.0,63504.0,99990.0
geolocation_lat,1000163.0,NaN,NaN,NaN,-21.176153,5.715866,-36.605374,-23.603546,-22.919377,-19.97962,45.065933
geolocation_lng,1000163.0,NaN,NaN,NaN,-46.390541,4.269748,-101.466766,-48.573172,-46.637879,-43.767709,121.105394
geolocation_city,1000163,8011,sao paulo,135800,NaN,NaN,NaN,NaN,NaN,NaN,NaN
geolocation_state,1000163,27,SP,404268,NaN,NaN,NaN,NaN,NaN,NaN,NaN



[5] MISSING VALUES


,Missing Count,Missing Percentage (%)
geolocation_zip_code_prefix,0,0.0
geolocation_lat,0,0.0
geolocation_lng,0,0.0
geolocation_city,0,0.0
geolocation_state,0,0.0



[6] DUPLICATED ROWS


,Duplicate Rows,Duplicate Percentage (%)
0,261831,26.18



[7] CARDINALITY


,Unique Values
geolocation_lng,717613
geolocation_lat,717360
geolocation_zip_code_prefix,19015
geolocation_city,8011
geolocation_state,27



DATASET: ORDER_ITEMS

[1] SAMPLE DATA (10 ROWS)


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
83939,beae8ee9872c468af6923d5d4e0ad6d8,1,ed6905f4a4a9728bc20da105c7099120,7901646fdd36a55f564ffaf2dbccaaf7,2017-09-06 17:05:39,465.52,18.33
1844,042c384c3edbe39a47895217eb5e90a4,1,b3cd491b3a347c8becc23fa5515440f9,b6d44737c043328708f6749c2dbe50bd,2017-05-09 14:02:33,145.00,28.20
42722,614c7adfc74985b5d111c1e38f0ad0cd,1,409ad0594dd6af287d870a566756dc10,d91fb3b7d041e83b64a00a3edfb37e4f,2018-08-01 12:30:38,175.00,19.33
21817,31d6ca721a02109e3f90202f01c62a4a,1,19944bc70a13c9b62e18a2bc7e913bb6,0bae85eb84b9fb3bd773911e89288d54,2018-03-15 00:48:50,24.95,19.32
49959,718b31ebf4b412db36cb480229bca6ab,1,e59dd207c69d86e890febadc796d1078,e9bc59e7b60fc3063eb2290deda4cced,2018-05-11 16:31:37,148.00,9.91
101917,e74c8d6962373b0c0e984e97fe2fa674,1,1b4dee3e58cb87cb5a0acc54ea53db9d,7142540dd4c91e2237acb7e911c4eba2,2017-06-02 13:06:09,99.90,29.10
96586,db2f0b731c53190c15527344d01f22f8,1,e6fd80a00ebb1a3e0d58a38555fbbf2c,f262cbc1c910c83959f849465454ddd3,2018-02-22 17:31:16,22.99,15.79
17016,26f48bf7cbe8a240db0273cdcfaba9b2,1,b27ff4b438c93e0dbc08bf543dbe31ed,5cf13accae3222c70a9cac40818ae839,2017-03-09 17:15:11,79.50,11.17
108603,f6d9acdd7ca0c1bc3caad5661a1be806,1,437c05a395e9e47f9762e677a7068ce7,bf84056e679dbe9c69929847a40e338f,2018-02-27 22:48:19,47.21,15.10
36319,525d105e2cd069520f58e5dc01013b3a,1,28b2775c30336108f0b6b4df4c7097cb,70849ca4f400aaabb62cb7462a6f1428,2018-05-21 19:55:22,16.99,13.47



[2] DATA SHAPE


,Rows,Columns
0,112650,7



[3] DATA TYPES


,Column,Data Type
order_id,order_id,object
order_item_id,order_item_id,int64
product_id,product_id,object
seller_id,seller_id,object
shipping_limit_date,shipping_limit_date,object
price,price,float64
freight_value,freight_value,float64



[4] DESCRIPTIVE STATISTICS


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
order_id,112650,98666,8272b63d03f5f79c56e9e4120aec44ef,21,NaN,NaN,NaN,NaN,NaN,NaN,NaN
order_item_id,112650.0,NaN,NaN,NaN,1.197834,0.705124,1.0,1.0,1.0,1.0,21.0
product_id,112650,32951,aca2eb7d00ea1a7b8ebd4e68314663af,527,NaN,NaN,NaN,NaN,NaN,NaN,NaN
seller_id,112650,3095,6560211a19b47992c3666cc44a7e94c0,2033,NaN,NaN,NaN,NaN,NaN,NaN,NaN
shipping_limit_date,112650,93318,2017-07-21 18:25:23,21,NaN,NaN,NaN,NaN,NaN,NaN,NaN
price,112650.0,NaN,NaN,NaN,120.653739,183.633928,0.85,39.9,74.99,134.9,6735.0
freight_value,112650.0,NaN,NaN,NaN,19.99032,15.806405,0.0,13.08,16.26,21.15,409.68



[5] MISSING VALUES


,Missing Count,Missing Percentage (%)
order_id,0,0.0
order_item_id,0,0.0
product_id,0,0.0
seller_id,0,0.0
shipping_limit_date,0,0.0
price,0,0.0
freight_value,0,0.0



[6] DUPLICATED ROWS


,Duplicate Rows,Duplicate Percentage (%)
0,0,0.0



[7] CARDINALITY


,Unique Values
order_id,98666
shipping_limit_date,93318
product_id,32951
freight_value,6999
price,5968
seller_id,3095
order_item_id,21



DATASET: PAYMENTS

[1] SAMPLE DATA (10 ROWS)


,order_id,payment_sequential,payment_type,payment_installments,payment_value
47094,2da878ecf6888b9ef86f601d9a8b9492,1,credit_card,2,115.45
22961,ee27c5db38b1e5eba4ca12bbe0cef731,1,credit_card,8,318.00
61794,ce14e551b308d863fc17c8c43f15d87b,1,credit_card,1,141.08
99493,53051fdaeb4854f9cbde7f3bec024317,1,credit_card,5,56.22
64279,9761e698e3d390c91d88364fb6a5ef11,1,credit_card,3,33.00
17883,b465b096e042571f13c2e9e8e15122a5,2,voucher,1,157.50
52478,556e5ad87434de71e0948b9bca0046f5,1,credit_card,1,56.22
15586,f50f12991ec8bcff03994381dcf9ce08,1,credit_card,1,28.62
66068,1a1b532e9fc48a702923b8ef014735e9,1,credit_card,1,139.24
55535,cefe70775d8821c8dcf166f843d38d1d,1,credit_card,1,75.16



[2] DATA SHAPE


,Rows,Columns
0,103886,5



[3] DATA TYPES


,Column,Data Type
order_id,order_id,object
payment_sequential,payment_sequential,int64
payment_type,payment_type,object
payment_installments,payment_installments,int64
payment_value,payment_value,float64



[4] DESCRIPTIVE STATISTICS


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
order_id,103886,99440,fa65dad1b0e818e3ccc5cb0e39231352,29,NaN,NaN,NaN,NaN,NaN,NaN,NaN
payment_sequential,103886.0,NaN,NaN,NaN,1.092679,0.706584,1.0,1.0,1.0,1.0,29.0
payment_type,103886,5,credit_card,76795,NaN,NaN,NaN,NaN,NaN,NaN,NaN
payment_installments,103886.0,NaN,NaN,NaN,2.853349,2.687051,0.0,1.0,1.0,4.0,24.0
payment_value,103886.0,NaN,NaN,NaN,154.10038,217.494064,0.0,56.79,100.0,171.8375,13664.08



[5] MISSING VALUES


,Missing Count,Missing Percentage (%)
order_id,0,0.0
payment_sequential,0,0.0
payment_type,0,0.0
payment_installments,0,0.0
payment_value,0,0.0



[6] DUPLICATED ROWS


,Duplicate Rows,Duplicate Percentage (%)
0,0,0.0



[7] CARDINALITY


,Unique Values
order_id,99440
payment_value,29077
payment_sequential,29
payment_installments,24
payment_type,5



DATASET: REVIEWS

[1] SAMPLE DATA (10 ROWS)


,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
60273,e32b690af8d0c370f6a0d5eb5f39ec55,cf3be801bb6918904774e23f8f5db5e1,5,SUPER RECOMENDO,"MUITO OBRIGADA, CHEGOU TUDO OK E VEIO ANTES DO...",2018-05-17 00:00:00,2018-05-22 14:57:43
37525,fee23d4e1207b57c75cc878b845f308f,4dcf81f76e3af4f8740d2b3c159a453d,3,NaN,Achei pequeno o relógio \r\nÉ melhor postar fo...,2018-04-21 00:00:00,2018-04-23 10:46:09
84437,f827a6cef42f7b9da3367fe73c975d60,fa3b5273908399cce287019297580550,5,NaN,"Recomendo, meu produto chegou bem antes do pra...",2017-12-22 00:00:00,2017-12-23 02:32:38
56155,35385a497b3eb7936ff743dbdef43625,612122563fa653c20d2ee959472a96c9,5,NaN,Só fico muito triste por não ter o cartão da l...,2017-10-25 00:00:00,2017-10-26 11:54:14
73414,063b9793774bd3df87799328d1f560d2,248b9e7139f9afa6227c51a42ecb915a,5,NaN,"Super indico o Site,... Entregaram bem antes d...",2018-04-12 00:00:00,2018-04-14 15:08:49
98508,f8af46f048f1a2a8bf849e14a085f176,8dcdc3f032fcbc54ae56d5a0f0c2f18c,4,NaN,NaN,2017-03-31 00:00:00,2017-04-01 13:41:53
98823,2db65dd06b38bcfa534230491ca89c96,27d948e8e3d9abb11648cfa05b1ce3f7,1,NaN,NaN,2017-11-02 00:00:00,2017-11-02 14:39:40
28838,76e025728461df638fd0047be9044b35,2f53a7c846aa70472520bd18b081ebbb,3,NaN,O produto apresentado na foto não condiz com o...,2017-09-21 00:00:00,2017-09-22 21:53:58
62613,c2ed33bcbe7d48bdb67bc5a31bdb442e,8f13ca9482f62f69781a1141ebad41e9,1,Não recomendo,Não recebi está demorando muito,2018-06-07 00:00:00,2018-06-07 13:35:32
70345,5de2541f4ef6cb873b10673a727f5eb1,989c69b80eadc79ccae5fe89c92a35ce,5,NaN,NaN,2018-06-27 00:00:00,2018-07-12 23:44:42



[2] DATA SHAPE


,Rows,Columns
0,99224,7



[3] DATA TYPES


,Column,Data Type
review_id,review_id,object
order_id,order_id,object
review_score,review_score,int64
review_comment_title,review_comment_title,object
review_comment_message,review_comment_message,object
review_creation_date,review_creation_date,object
review_answer_timestamp,review_answer_timestamp,object



[4] DESCRIPTIVE STATISTICS


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
review_id,99224,98410,08528f70f579f0c830189efc523d2182,3,NaN,NaN,NaN,NaN,NaN,NaN,NaN
order_id,99224,98673,df56136b8031ecd28e200bb18e6ddb2e,3,NaN,NaN,NaN,NaN,NaN,NaN,NaN
review_score,99224.0,NaN,NaN,NaN,4.086421,1.347579,1.0,4.0,5.0,5.0,5.0
review_comment_title,11568,4527,Recomendo,423,NaN,NaN,NaN,NaN,NaN,NaN,NaN
review_comment_message,40977,36159,Muito bom,230,NaN,NaN,NaN,NaN,NaN,NaN,NaN
review_creation_date,99224,636,2017-12-19 00:00:00,463,NaN,NaN,NaN,NaN,NaN,NaN,NaN
review_answer_timestamp,99224,98248,2017-06-15 23:21:05,4,NaN,NaN,NaN,NaN,NaN,NaN,NaN



[5] MISSING VALUES


,Missing Count,Missing Percentage (%)
review_comment_title,87656,88.34
review_comment_message,58247,58.70
review_id,0,0.00
review_score,0,0.00
order_id,0,0.00
review_creation_date,0,0.00
review_answer_timestamp,0,0.00



[6] DUPLICATED ROWS


,Duplicate Rows,Duplicate Percentage (%)
0,0,0.0



[7] CARDINALITY


,Unique Values
order_id,98673
review_id,98410
review_answer_timestamp,98248
review_comment_message,36159
review_comment_title,4527
review_creation_date,636
review_score,5



DATASET: ORDERS

[1] SAMPLE DATA (10 ROWS)


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
55335,88aef6eed6c5cf608b6e57b402ca30ca,7ac6d2ac9e318ec647bf58a9049707b2,delivered,2018-08-22 13:02:19,2018-08-22 13:10:23,2018-08-28 05:41:00,2018-08-29 04:52:45,2018-08-31 00:00:00
25338,7a8f0c1255003a0de7fd484dbab75fdb,04b48387bf71a20a7b51d540080e50f2,delivered,2018-07-05 12:10:32,2018-07-06 02:55:12,2018-07-13 13:38:00,2018-07-23 15:14:55,2018-08-13 00:00:00
81091,c4c55b747930acd84e6c499e461852df,fa3e77981c620b52d92ee2d390543f63,delivered,2017-11-29 10:29:45,2017-11-29 10:37:29,2017-12-05 20:17:51,2017-12-08 14:23:11,2017-12-13 00:00:00
89852,d89ff75da06f798ea64c91a496386314,cb93a9fab20aef775dac6e14672ce6df,delivered,2018-01-09 18:55:23,2018-01-10 10:31:43,2018-01-10 22:21:01,2018-02-02 16:58:44,2018-02-06 00:00:00
57030,c03a1a47065000233e19e37a01eca310,4dd50db22d8966118f82f5157168f4c6,delivered,2017-06-18 00:06:49,2017-06-18 00:24:57,2017-06-19 15:06:52,2017-06-27 09:10:47,2017-07-07 00:00:00
67173,d80bc8900db383324d145cfb1c1aa854,1de6c6c911ceda2be652cd0db03d930e,delivered,2018-05-26 07:48:45,2018-05-26 08:19:52,2018-05-29 14:10:00,2018-06-07 18:51:46,2018-07-04 00:00:00
63523,95606e8113c2e8087c09cffe458fbf82,ac3d7f758530a079398af1d277f48d22,delivered,2017-05-16 10:15:30,2017-05-16 10:25:16,2017-05-17 09:32:15,2017-05-25 14:33:13,2017-06-08 00:00:00
89731,660a8ca012d4806b05c2a9b705570f62,6f728d2ab5c661f6b32499e7e1bd3261,delivered,2018-03-29 10:25:37,2018-03-30 03:15:51,2018-04-02 17:37:53,2018-04-19 21:34:29,2018-05-08 00:00:00
76185,acaaf1ded8aad100b7760b0ba16aad86,aad03b10f771077832a29862911f79a1,delivered,2017-09-09 13:21:51,2017-09-12 04:26:08,2017-09-14 18:33:31,2017-09-21 20:58:59,2017-10-03 00:00:00
10498,9c7786ec8d2394cbee42bba833f7c537,9a935364cca30059364ef4469083e8f8,delivered,2017-04-27 23:45:10,2017-04-28 11:46:07,2017-04-28 10:57:41,2017-05-05 15:49:46,2017-05-30 00:00:00



[2] DATA SHAPE


,Rows,Columns
0,99441,8



[3] DATA TYPES


,Column,Data Type
order_id,order_id,object
customer_id,customer_id,object
order_status,order_status,object
order_purchase_timestamp,order_purchase_timestamp,object
order_approved_at,order_approved_at,object
order_delivered_carrier_date,order_delivered_carrier_date,object
order_delivered_customer_date,order_delivered_customer_date,object
order_estimated_delivery_date,order_estimated_delivery_date,object



[4] DESCRIPTIVE STATISTICS


,count,unique,top,freq
order_id,99441,99441,66dea50a8b16d9b4dee7af250b4be1a5,1
customer_id,99441,99441,edb027a75a1449115f6b43211ae02a24,1
order_status,99441,8,delivered,96478
order_purchase_timestamp,99441,98875,2018-08-02 12:05:26,3
order_approved_at,99281,90733,2018-02-27 04:31:10,9
order_delivered_carrier_date,97658,81018,2018-05-09 15:48:00,47
order_delivered_customer_date,96476,95664,2018-05-08 19:36:48,3
order_estimated_delivery_date,99441,459,2017-12-20 00:00:00,522



[5] MISSING VALUES


,Missing Count,Missing Percentage (%)
order_delivered_customer_date,2965,2.98
order_delivered_carrier_date,1783,1.79
order_approved_at,160,0.16
order_id,0,0.00
order_purchase_timestamp,0,0.00
order_status,0,0.00
customer_id,0,0.00
order_estimated_delivery_date,0,0.00



[6] DUPLICATED ROWS


,Duplicate Rows,Duplicate Percentage (%)
0,0,0.0



[7] CARDINALITY


,Unique Values
order_id,99441
customer_id,99441
order_purchase_timestamp,98875
order_delivered_customer_date,95664
order_approved_at,90733
order_delivered_carrier_date,81018
order_estimated_delivery_date,459
order_status,8



DATASET: PRODUCTS

[1] SAMPLE DATA (10 ROWS)


,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
19421,15f17270133dfedc0e1064782a60e374,cool_stuff,51.0,305.0,3.0,6000.0,52.0,44.0,27.0
13132,65d7a1724c5c7fd77ac36c80f241ba6a,automotivo,59.0,211.0,4.0,600.0,25.0,10.0,15.0
701,606bfa78f6209099f25529bb1211522f,moveis_decoracao,40.0,241.0,2.0,4000.0,52.0,6.0,52.0
9649,9766ad53f42ee034fbc0e29c85fe626c,cama_mesa_banho,51.0,385.0,2.0,2500.0,47.0,21.0,41.0
24467,41ce18e06273b50422ebf6149066fa23,papelaria,36.0,41.0,1.0,100.0,16.0,5.0,16.0
18310,56b81eaffe3c2e7e54f3bdf2a935164f,moveis_decoracao,56.0,296.0,3.0,1400.0,100.0,7.0,15.0
1876,5c15e8f223a8bf0d7858c403ddc1b0ed,moveis_decoracao,63.0,699.0,1.0,650.0,45.0,15.0,35.0
32948,9a7c6041fa9592d9d9ef6cfe62a71f8c,cama_mesa_banho,50.0,799.0,1.0,1400.0,27.0,7.0,27.0
19605,345f6c48d7dd28adc3242eafba3592ce,utilidades_domesticas,55.0,1725.0,1.0,825.0,26.0,15.0,22.0
1947,09e5556011177a1f20244a4fcdde2752,moveis_sala,52.0,400.0,3.0,6100.0,50.0,50.0,11.0



[2] DATA SHAPE


,Rows,Columns
0,32951,9



[3] DATA TYPES


,Column,Data Type
product_id,product_id,object
product_category_name,product_category_name,object
product_name_lenght,product_name_lenght,float64
product_description_lenght,product_description_lenght,float64
product_photos_qty,product_photos_qty,float64
product_weight_g,product_weight_g,float64
product_length_cm,product_length_cm,float64
product_height_cm,product_height_cm,float64
product_width_cm,product_width_cm,float64



[4] DESCRIPTIVE STATISTICS


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
product_id,32951,32951,106392145fca363410d287a815be6de4,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
product_category_name,32341,73,cama_mesa_banho,3029,NaN,NaN,NaN,NaN,NaN,NaN,NaN
product_name_lenght,32341.0,NaN,NaN,NaN,48.476949,10.245741,5.0,42.0,51.0,57.0,76.0
product_description_lenght,32341.0,NaN,NaN,NaN,771.495285,635.115225,4.0,339.0,595.0,972.0,3992.0
product_photos_qty,32341.0,NaN,NaN,NaN,2.188986,1.736766,1.0,1.0,1.0,3.0,20.0
product_weight_g,32949.0,NaN,NaN,NaN,2276.472488,4282.038731,0.0,300.0,700.0,1900.0,40425.0
product_length_cm,32949.0,NaN,NaN,NaN,30.815078,16.914458,7.0,18.0,25.0,38.0,105.0
product_height_cm,32949.0,NaN,NaN,NaN,16.937661,13.637554,2.0,8.0,13.0,21.0,105.0
product_width_cm,32949.0,NaN,NaN,NaN,23.196728,12.079047,6.0,15.0,20.0,30.0,118.0



[5] MISSING VALUES


,Missing Count,Missing Percentage (%)
product_category_name,610,1.85
product_description_lenght,610,1.85
product_name_lenght,610,1.85
product_photos_qty,610,1.85
product_weight_g,2,0.01
product_height_cm,2,0.01
product_length_cm,2,0.01
product_width_cm,2,0.01
product_id,0,0.00



[6] DUPLICATED ROWS


,Duplicate Rows,Duplicate Percentage (%)
0,0,0.0



[7] CARDINALITY


,Unique Values
product_id,32951
product_description_lenght,2960
product_weight_g,2204
product_height_cm,102
product_length_cm,99
product_width_cm,95
product_category_name,73
product_name_lenght,66
product_photos_qty,19



DATASET: SELLERS

[1] SAMPLE DATA (10 ROWS)


,seller_id,seller_zip_code_prefix,seller_city,seller_state
486,70f84699be677ec988d1fa824aef7d90,35700,sete lagoas,MG
765,ba90964cff9b9e0e6f32b23b82465f7b,7140,guarulhos,SP
313,a7c88262832dfc683781d6a421edefae,8710,mogi das cruzes,SP
1949,77714512e8d4dda9c234da8e3faf16e4,7056,guarulhos,SP
2273,4c03b9dd4c11ee2cb35c96c49efc9420,13232,campo limpo paulista,SP
1211,5305693ffae2d3463377b1f6fe67b15a,16300,penapolis,SP
1570,21c62b998a043ebe31161d38f84929fc,60110,fortaleza,CE
2129,0adac9fbd9a2b63cccaac4f8756c1ca8,13290,louveira,SP
1816,272f092de69afedd4d2969440b37f18f,3265,sao paulo,SP
116,37be5a7c751166fbc5f8ccba4119e043,4248,sao paulo,SP



[2] DATA SHAPE


,Rows,Columns
0,3095,4



[3] DATA TYPES


,Column,Data Type
seller_id,seller_id,object
seller_zip_code_prefix,seller_zip_code_prefix,int64
seller_city,seller_city,object
seller_state,seller_state,object



[4] DESCRIPTIVE STATISTICS


,count,unique,top,freq,mean,std,min,25%,50%,75%,max
seller_id,3095,3095,9e25199f6ef7e7c347120ff175652c3b,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
seller_zip_code_prefix,3095.0,NaN,NaN,NaN,32291.059451,32713.45383,1001.0,7093.5,14940.0,64552.5,99730.0
seller_city,3095,611,sao paulo,694,NaN,NaN,NaN,NaN,NaN,NaN,NaN
seller_state,3095,23,SP,1849,NaN,NaN,NaN,NaN,NaN,NaN,NaN



[5] MISSING VALUES


,Missing Count,Missing Percentage (%)
seller_id,0,0.0
seller_zip_code_prefix,0,0.0
seller_city,0,0.0
seller_state,0,0.0



[6] DUPLICATED ROWS


,Duplicate Rows,Duplicate Percentage (%)
0,0,0.0



[7] CARDINALITY


,Unique Values
seller_id,3095
seller_zip_code_prefix,2246
seller_city,611
seller_state,23



DATASET: CATEGORY_TRANSLATION

[1] SAMPLE DATA (10 ROWS)


,product_category_name,product_category_name_english
31,market_place,market_place
32,eletronicos,electronics
40,casa_construcao,home_construction
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
59,cine_foto,cine_photo
8,telefonia,telephony
51,fashion_esporte,fashion_sport
15,telefonia_fixa,fixed_telephony
57,livros_importados,books_imported



[2] DATA SHAPE


,Rows,Columns
0,71,2



[3] DATA TYPES


,Column,Data Type
product_category_name,product_category_name,object
product_category_name_english,product_category_name_english,object



[4] DESCRIPTIVE STATISTICS


,count,unique,top,freq
product_category_name,71,71,beleza_saude,1
product_category_name_english,71,71,health_beauty,1



[5] MISSING VALUES


,Missing Count,Missing Percentage (%)
product_category_name,0,0.0
product_category_name_english,0,0.0



[6] DUPLICATED ROWS


,Duplicate Rows,Duplicate Percentage (%)
0,0,0.0



[7] CARDINALITY


,Unique Values
product_category_name,71
product_category_name_english,71


In [25]:
hotel_df = dfs["hotel_bookings"]
customers_df = dfs["customers"]
geolocation_df = dfs["geolocation"]
order_items_df = dfs["order_items"]
payments_df = dfs["payments"]
reviews_df = dfs["reviews"]
orders_df = dfs["orders"]
products_df = dfs["products"]
sellers_df = dfs["sellers"]
category_df = dfs["category_translation"]

## Handling Missing Values in Hotel Booking Dataset

Trong quá trình khám phá dữ liệu, bảng `hotel_bookings` xuất hiện một số cột chứa giá trị thiếu (missing values). Các phương pháp xử lý được lựa chọn dựa trên ý nghĩa nghiệp vụ của từng thuộc tính:

* **company**: Cột này có tỷ lệ thiếu rất cao (94.31%), do đó được loại bỏ khỏi tập dữ liệu vì không mang lại nhiều giá trị cho phân tích và mô hình học máy.
* **agent**: Giá trị thiếu được thay thế bằng `0`, biểu thị khách hàng không đặt phòng thông qua đại lý du lịch.
* **country**: Giá trị thiếu được thay thế bằng `"Unknown"` nhằm giữ lại toàn bộ bản ghi và phản ánh rõ các trường hợp không xác định quốc gia.
* **children**: Giá trị thiếu được thay thế bằng `0`, giả định rằng khách hàng không đi cùng trẻ em trong các trường hợp dữ liệu bị thiếu.

Sau khi xử lý, tập dữ liệu không còn giá trị thiếu ở các cột trên và sẵn sàng cho các bước phân tích và xây dựng mô hình tiếp theo.


In [26]:

hotel_df.drop(columns=["company"], inplace=True)

hotel_df["agent"] = hotel_df["agent"].fillna(0)

hotel_df["country"] = hotel_df["country"].fillna("Unknown")

hotel_df["children"] = hotel_df["children"].fillna(0)

## Handling Missing Values in Review Dataset

Bảng `olist_order_reviews_dataset` chứa hai cột văn bản là `review_comment_title` và `review_comment_message` với tỷ lệ giá trị thiếu khá cao. Tuy nhiên, các giá trị thiếu này không phải là lỗi dữ liệu mà phản ánh trường hợp khách hàng chỉ đánh giá bằng số sao mà không để lại nhận xét bằng văn bản.

Do đó, thay vì xóa các bản ghi hoặc loại bỏ cột, các giá trị thiếu được thay thế bằng:

* `No Title` đối với cột `review_comment_title`
* `No Message` đối với cột `review_comment_message`

Cách xử lý này giúp bảo toàn toàn bộ dữ liệu đánh giá và hỗ trợ cho các bước phân tích văn bản hoặc xây dựng mô hình học máy trong tương lai.


In [27]:
reviews_df["review_comment_title"] = (
    reviews_df["review_comment_title"]
    .fillna("No Title")
)

reviews_df["review_comment_message"] = (
    reviews_df["review_comment_message"]
    .fillna("No Message")
)


## Converting Date Columns to Datetime

Các cột thời gian trong bảng `orders` được chuyển từ kiểu dữ liệu `object` sang `datetime` để hỗ trợ việc phân tích dữ liệu theo thời gian, tính toán thời gian giao hàng và xây dựng đặc trưng cho các mô hình học máy.

Các cột được chuyển đổi bao gồm:

* order_purchase_timestamp
* order_approved_at
* order_delivered_carrier_date
* order_delivered_customer_date
* order_estimated_delivery_date

Trong quá trình chuyển đổi, các giá trị không hợp lệ sẽ được tự động chuyển thành `NaT` bằng tham số `errors='coerce'`.


## Handling Missing Values in Order Date Columns

Trong bảng `orders`, ba cột `order_approved_at`, `order_delivered_carrier_date` và `order_delivered_customer_date` xuất hiện một số giá trị thiếu. Sau khi kiểm tra mối quan hệ với thuộc tính `order_status`, nhận thấy các giá trị thiếu này chủ yếu xuất hiện ở những đơn hàng có trạng thái như `canceled`, `unavailable`, `created` hoặc các đơn hàng chưa hoàn tất quy trình xử lý.

Do đó, các giá trị thiếu này không phải là lỗi dữ liệu mà phản ánh đúng trạng thái thực tế của đơn hàng trong hệ thống:

* `order_approved_at`: Đơn hàng chưa được xác nhận thanh toán hoặc bị hủy trước khi hoàn tất quá trình phê duyệt.
* `order_delivered_carrier_date`: Đơn hàng chưa được bàn giao cho đơn vị vận chuyển.
* `order_delivered_customer_date`: Đơn hàng chưa được giao thành công cho khách hàng.

Việc thay thế các giá trị thiếu bằng một ngày giả định có thể làm sai lệch bản chất nghiệp vụ của dữ liệu và ảnh hưởng đến các phân tích liên quan đến thời gian xử lý, vận chuyển và giao hàng. Vì vậy, các giá trị thiếu trong ba cột này được giữ nguyên để đảm bảo tính chính xác và toàn vẹn của dữ liệu.



In [28]:
date_columns = [
    'order_purchase_timestamp',
    'order_approved_at',
    'order_delivered_carrier_date',
    'order_delivered_customer_date',
    'order_estimated_delivery_date'
]

orders_df[date_columns] = orders_df[date_columns].apply(
    pd.to_datetime,
    errors='coerce'
)

display(
    pd.DataFrame({
        "Column": date_columns,
        "Data Type": orders_df[date_columns].dtypes.astype(str)
    })
)

,Column,Data Type
order_purchase_timestamp,order_purchase_timestamp,datetime64[ns]
order_approved_at,order_approved_at,datetime64[ns]
order_delivered_carrier_date,order_delivered_carrier_date,datetime64[ns]
order_delivered_customer_date,order_delivered_customer_date,datetime64[ns]
order_estimated_delivery_date,order_estimated_delivery_date,datetime64[ns]


In [29]:
orders_df.isnull().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

## Handling Missing Values in Product Dataset

Các giá trị thiếu trong bảng sản phẩm được xử lý dựa trên bản chất của từng thuộc tính.

* `product_category_name` được thay thế bằng `"Unknown Category"` nhằm biểu diễn các sản phẩm chưa xác định được danh mục.
* `product_name_lenght`, `product_description_lenght` và `product_photos_qty` được thay thế bằng giá trị `0`, phản ánh trường hợp sản phẩm không có tên, không có mô tả hoặc không có hình ảnh.
* Các thuộc tính vật lý của sản phẩm gồm `product_weight_g`, `product_length_cm`, `product_height_cm` và `product_width_cm` được thay thế bằng giá trị trung vị (Median). Phương pháp này giúp giảm ảnh hưởng của các giá trị ngoại lai và duy trì phân bố dữ liệu gần với thực tế.

Sau khi xử lý, tập dữ liệu không còn giá trị thiếu và sẵn sàng cho các bước phân tích và xây dựng mô hình học máy.


In [30]:
products_df['product_category_name'] = (
    products_df['product_category_name']
    .fillna('Unknown Category')
)


products_df['product_name_lenght'] = (
    products_df['product_name_lenght']
    .fillna(0)
)

products_df['product_description_lenght'] = (
    products_df['product_description_lenght']
    .fillna(0)
)

products_df['product_photos_qty'] = (
    products_df['product_photos_qty']
    .fillna(0)
)

physical_cols = [
    'product_weight_g',
    'product_length_cm',
    'product_height_cm',
    'product_width_cm'
]

for col in physical_cols:
    products_df[col] = products_df[col].fillna(
        products_df[col].median()
    )

In [31]:
for name, df in {
    "hotel": hotel_df,
    "reviews": reviews_df,
    "orders": orders_df,
    "products": products_df
}.items():

    print(f"\n{name.upper()}")
    print(df.isnull().sum()[df.isnull().sum() > 0])


HOTEL
Series([], dtype: int64)

REVIEWS
Series([], dtype: int64)

ORDERS
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
dtype: int64

PRODUCTS
Series([], dtype: int64)


In [33]:
for name, df in dfs.items():
    print(f"{name}: {df.duplicated().sum()} duplicate rows")

hotel_bookings: 32001 duplicate rows
customers: 0 duplicate rows
geolocation: 261831 duplicate rows
order_items: 0 duplicate rows
payments: 0 duplicate rows
reviews: 0 duplicate rows
orders: 0 duplicate rows
products: 0 duplicate rows
sellers: 0 duplicate rows
category_translation: 0 duplicate rows
